storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java

```diff
@@ -0,0 +1,390 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.storm.daemon;
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.sun.net.httpserver.HttpsServer;
+import com.sun.org.apache.bcel.internal.generic.ARRAYLENGTH;
+import org.apache.commons.lang.StringUtils;
+import org.apache.storm.Config;
+import org.apache.storm.daemon.metrics.MetricsUtils;
+import org.apache.storm.daemon.metrics.reporters.PreparableReporter;
+import org.apache.storm.generated.*;
+import org.apache.storm.logging.ThriftAccessLogger;
+import org.apache.storm.security.auth.*;
+import org.apache.storm.security.auth.authorizer.DRPCAuthorizerBase;
+import org.apache.storm.ui.FilterConfiguration;
+import org.apache.storm.ui.IConfigurator;
+import org.apache.storm.ui.UIHelpers;
+import org.apache.storm.utils.ConfigUtils;
+import org.apache.storm.utils.Time;
+import org.apache.storm.utils.Utils;
+import org.apache.storm.utils.VersionInfo;
+import org.apache.thrift.TException;
+import org.eclipse.jetty.server.Server;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.Servlet;
+import java.security.Principal;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class DrpcServer implements DistributedRPC.Iface, DistributedRPCInvocations.Iface, Shutdownable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DrpcServer.class);
+    private final Long timeoutCheckSecs = 5L;
+
+    private Map conf;
+
+    private ThriftServer handlerServer;
+    private ThriftServer invokeServer;
+    private IHttpCredentialsPlugin httpCredsHandler;
+
+    private Thread clearThread;
+
+    private IAuthorizer authorizer;
+
+    // To be removed after porting drpc.clj
+    private Servlet httpServlet;
+
+    private AtomicInteger ctr = new AtomicInteger(0);
+    private ConcurrentHashMap<String, Semaphore> idtoSem = new ConcurrentHashMap<String, Semaphore>();
+    private ConcurrentHashMap<String, Object> idtoResult = new ConcurrentHashMap<String, Object>();
+    private ConcurrentHashMap<String, Integer> idtoStart = new ConcurrentHashMap<String, Integer>();
+    private ConcurrentHashMap<String, String> idtoFunction = new ConcurrentHashMap<String, String>();
+    private ConcurrentHashMap<String, DRPCRequest> idtoRequest = new ConcurrentHashMap<String, DRPCRequest>();
+    private ConcurrentHashMap<String, ConcurrentLinkedQueue<DRPCRequest>> requestQueues = new ConcurrentHashMap<String, ConcurrentLinkedQueue<DRPCRequest>>();
```

Comment:
This is a perfect place to actually improve things to make them more object oriented.  The key to almost all of these maps is the same thing.  Why not remove all of the special ordering that we have to do on setup and cleanup.

```
private static class InternalRequest {
    public final Semaphore sem;
    public final long startTimeSecs;
    public final string function;
    public final DRPCRequest request;
    public volatile Object result;

    public InternalRequest(String function, DRPCRequest request) {
        sem = new Semaphore(0);
        startTimeSecs = Time.currentTimeSecs();
        this.function = function;
        this.request = request;
    }
}

private ConcurrentHashMap<String, InternalRequest> outstandingRequests = new ConcurrentHashMap<>();
```